src/o365/aad/commands/approleassignment/approleassignment-add.ts

```diff
@@ -0,0 +1,298 @@
+import commands from '../../commands';
+import GlobalOptions from '../../../../GlobalOptions';
+import request from '../../../../request';
+import {
+  CommandOption,
+  CommandValidate
+} from '../../../../Command';
+import AadCommand from '../../../base/AadCommand';
+import Utils from '../../../../Utils';
+import { ServicePrincipal } from './ServicePrincipal';
+
+const vorpal: Vorpal = require('../../../../vorpal-init');
+
+interface AppRole {
+  objectId: string;
+  value: string;
+  resourceId: string;
+}
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  appId?: string;
+  objectId?: string;
+  displayName?: string;
+  resource: string;
+  scope: string;
+}
+
+class AadAppRoleAssignmentAddCommand extends AadCommand {
+
+  public get name(): string {
+    return commands.APPROLEASSIGNMENT_ADD;
+  }
+
+  public get description(): string {
+    return 'Adds service principal permissions also known as scopes and app role assignments for specified Azure AD application registration';
+  }
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.appId = typeof args.options.appId !== 'undefined';
+    telemetryProps.objectId = typeof args.options.objectId !== 'undefined';
+    telemetryProps.displayName = typeof args.options.displayName !== 'undefined';
+    return telemetryProps;
+  }
+
+  public commandAction(cmd: CommandInstance, args: CommandArgs, cb: () => void): void {
+    let objectId: string = '';
+    let queryFilter: string = '';
+    if (args.options.appId) {
+      queryFilter = `$filter=appId eq '${encodeURIComponent(args.options.appId)}'`;
+    } else if (args.options.objectId) {
+      queryFilter = `$filter=objectId eq '${encodeURIComponent(args.options.objectId)}'`;
+    } else {
+      queryFilter = `$filter=displayName eq '${encodeURIComponent(args.options.displayName as string)}'`;
+    }
+
+    const getServicePrinciplesRequestOptions: any = {
+      url: `${this.resource}/myorganization/servicePrincipals?api-version=1.6&${queryFilter}`,
+      headers: {
+        accept: 'application/json;odata=nometadata;streaming=false'
+      },
+      json: true
+    };
+
+    request
+      .get<{ value: ServicePrincipal[] }>(getServicePrinciplesRequestOptions)
+      .then((servicePrincipalResult: { value: ServicePrincipal[] }): Promise<{ value: ServicePrincipal[] }> => {
+
+        objectId = servicePrincipalResult.value[0].objectId;
```

Comment:
While the `objectId` and `appId` are unique, it's possible to have multiple service principals with the same name. In such case, we might end up adding permissions to the wrong principal. I'd like to suggest, that if we retrieved multiple service principals, we return an error asking the user to disambiguate to which principal we should permissions to avoid incorrect assignments.

Reply:
Good point. I will send error in case of more than one items in this array.

Reply:
@waldekmastykarz , I made the change and it is in a separate commit for you to review https://github.com/pnp/office365-cli/pull/1584/commits/bf8ac83cef2963a6e40ee670906de2aa9f65fc7a.

Please let me know and I will squash the changes and update this branch if needed.

Reply:
Looks good! I'll have another pass at it and squash & merge if all works as expected! Thanks for a quick turnaround! 👏